assets/javascript/apps/pipeline/nodes/widgets.tsx

```diff
@@ -1101,3 +1104,219 @@ function BuiltInToolsWidget(props: WidgetParams) {
+        </form>
+
+        <div className="grow h-full w-full flex flex-col">
+          <h4 className="mb-4 font-bold text-lg capitalize">{label}</h4>
+
+          <CodeMirror
+            value={value}
+            onChange={onChange}
+            height="100%"
+            theme={isDarkMode ? githubDark : githubLight}
+            extensions={[
+              autocompletion({
+                override: [textEditorVarCompletions(autocomplete_vars_list)],
+                activateOnTyping: true,
+              }),
+              highlightAutoCompleteVars(autocomplete_vars_list),
+              autocompleteVarTheme,
+              EditorView.lineWrapping,
+            ]}
+            basicSetup={{
+              lineNumbers: true,
+              tabSize: 2,
+              indentOnInput: true,
+            }}
+          />
+        </div>
+
+        {inputError && <div className="text-red-500">{inputError}</div>}
+      </div>
+      <form method="dialog" className="modal-backdrop">
+        <button>close</button>
+      </form>
+    </dialog>
+  );
+}
+
+function textEditorVarCompletions(autocomplete_vars_list: string[]) {
+  return (context: CompletionContext) => {
+    const word = context.matchBefore(/[a-zA-Z0-9._[\]"]*$/);
+    if (!word || (word.from === word.to && !context.explicit)) return null;
+
+    return {
+      from: word.from,
+      options: autocomplete_vars_list.map((v) => ({
+        label: v,
+        type: "variable",
+        info: `Insert {${v}}`,
+        apply: (view: EditorView, completion: Completion, from: number, to: number) => {
+          const beforeText = view.state.doc.sliceString(from - 1, from);
+          const insertText =
+            beforeText === "{" ? `${completion.label}` : `{${completion.label}}`;
+          view.dispatch({
+            changes: { from, to, insert: insertText },
+          });
+        },
+      })),
+    };
+  };
+}
+// highlight auto complete words. valid - blue, invalid - red
+function highlightAutoCompleteVars(autocomplete_vars_list: string[]) {
+  return ViewPlugin.fromClass(
+    class {
+      decorations: DecorationSet;
+
+      constructor(view: EditorView) {
+        this.decorations = this.buildDecorations(view);
+      }
+
+      update(update: ViewUpdate) {
+        if (update.docChanged || update.viewportChanged) {
+          this.decorations = this.buildDecorations(update.view);
+        }
+      }
+
+      buildDecorations(view: EditorView) {
+        const widgets: any[] = [];
+        const text = view.state.doc.toString();
+        const regex = /\{([a-zA-Z0-9._[\]"]+)\}/g;
+        let match;
+        while ((match = regex.exec(text)) !== null) {
+          const varName = match[1];
+          const from = match.index;
+          const to = from + match[0].length;
+
+          const deco = Decoration.mark({
+            class: autocomplete_vars_list.includes(varName)
+              ? "autocomplete-var-valid"
+              : "autocomplete-var-invalid",
```

Comment:
This should also handle cases where we are using dot notiation e.g. `{session_state.category}` or `{participant_data.schedules[0].name}`.

See https://docs.openchatstudio.com/concepts/prompt_variables/?h=prompt#nested-data-access